{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"run.ipynb","provenance":[],"collapsed_sections":[],"mount_file_id":"1PnKupvPnGQZPwdxJWi3t_u6AL0mMXasp","authorship_tag":"ABX9TyOdZHyBU24YJi0JZBZCm7Jn"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"},"accelerator":"GPU"},"cells":[{"cell_type":"code","source":["from google.colab import drive\n","drive.mount('/content/drive')"],"metadata":{"id":"aBzFSwlc97k2","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1642517204022,"user_tz":-60,"elapsed":1674,"user":{"displayName":"Francisco Pereira","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14Gju-L5-ouxVHtbPGNLO-WBYoOaPx1t8TQtxAGb4kA=s64","userId":"16300594535934500651"}},"outputId":"3d411063-1bce-4b67-aac6-7ee07d6bf9f9"},"execution_count":1,"outputs":[{"output_type":"stream","name":"stdout","text":["Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount(\"/content/drive\", force_remount=True).\n"]}]},{"cell_type":"code","source":["!pip install -r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt"],"metadata":{"id":"OQeHZ_Re4mwt","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1642517206668,"user_tz":-60,"elapsed":2654,"user":{"displayName":"Francisco Pereira","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14Gju-L5-ouxVHtbPGNLO-WBYoOaPx1t8TQtxAGb4kA=s64","userId":"16300594535934500651"}},"outputId":"fa3a01da-72bd-44a8-9f1b-21ef78af21ee"},"execution_count":2,"outputs":[{"output_type":"stream","name":"stdout","text":["Requirement already satisfied: numpy in /usr/local/lib/python3.7/dist-packages (from -r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 1)) (1.19.5)\n","Requirement already satisfied: tensorflow in /usr/local/lib/python3.7/dist-packages (from -r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (2.7.0)\n","Requirement already satisfied: keras in /usr/local/lib/python3.7/dist-packages (from -r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 3)) (2.7.0)\n","Requirement already satisfied: pandas in /usr/local/lib/python3.7/dist-packages (from -r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 4)) (1.1.5)\n","Requirement already satisfied: nltk in /usr/local/lib/python3.7/dist-packages (from -r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 5)) (3.2.5)\n","Requirement already satisfied: astunparse>=1.6.0 in /usr/local/lib/python3.7/dist-packages (from tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (1.6.3)\n","Requirement already satisfied: flatbuffers<3.0,>=1.12 in /usr/local/lib/python3.7/dist-packages (from tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (2.0)\n","Requirement already satisfied: google-pasta>=0.1.1 in /usr/local/lib/python3.7/dist-packages (from tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (0.2.0)\n","Requirement already satisfied: opt-einsum>=2.3.2 in /usr/local/lib/python3.7/dist-packages (from tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (3.3.0)\n","Requirement already satisfied: grpcio<2.0,>=1.24.3 in /usr/local/lib/python3.7/dist-packages (from tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (1.43.0)\n","Requirement already satisfied: keras-preprocessing>=1.1.1 in /usr/local/lib/python3.7/dist-packages (from tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (1.1.2)\n","Requirement already satisfied: termcolor>=1.1.0 in /usr/local/lib/python3.7/dist-packages (from tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (1.1.0)\n","Requirement already satisfied: h5py>=2.9.0 in /usr/local/lib/python3.7/dist-packages (from tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (3.1.0)\n","Requirement already satisfied: tensorflow-estimator<2.8,~=2.7.0rc0 in /usr/local/lib/python3.7/dist-packages (from tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (2.7.0)\n","Requirement already satisfied: libclang>=9.0.1 in /usr/local/lib/python3.7/dist-packages (from tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (12.0.0)\n","Requirement already satisfied: tensorflow-io-gcs-filesystem>=0.21.0 in /usr/local/lib/python3.7/dist-packages (from tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (0.23.1)\n","Requirement already satisfied: protobuf>=3.9.2 in /usr/local/lib/python3.7/dist-packages (from tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (3.17.3)\n","Requirement already satisfied: gast<0.5.0,>=0.2.1 in /usr/local/lib/python3.7/dist-packages (from tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (0.4.0)\n","Requirement already satisfied: tensorboard~=2.6 in /usr/local/lib/python3.7/dist-packages (from tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (2.7.0)\n","Requirement already satisfied: absl-py>=0.4.0 in /usr/local/lib/python3.7/dist-packages (from tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (0.12.0)\n","Requirement already satisfied: six>=1.12.0 in /usr/local/lib/python3.7/dist-packages (from tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (1.15.0)\n","Requirement already satisfied: wrapt>=1.11.0 in /usr/local/lib/python3.7/dist-packages (from tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (1.13.3)\n","Requirement already satisfied: typing-extensions>=3.6.6 in /usr/local/lib/python3.7/dist-packages (from tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (3.10.0.2)\n","Requirement already satisfied: wheel<1.0,>=0.32.0 in /usr/local/lib/python3.7/dist-packages (from tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (0.37.1)\n","Requirement already satisfied: cached-property in /usr/local/lib/python3.7/dist-packages (from h5py>=2.9.0->tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (1.5.2)\n","Requirement already satisfied: werkzeug>=0.11.15 in /usr/local/lib/python3.7/dist-packages (from tensorboard~=2.6->tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (1.0.1)\n","Requirement already satisfied: requests<3,>=2.21.0 in /usr/local/lib/python3.7/dist-packages (from tensorboard~=2.6->tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (2.23.0)\n","Requirement already satisfied: tensorboard-plugin-wit>=1.6.0 in /usr/local/lib/python3.7/dist-packages (from tensorboard~=2.6->tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (1.8.1)\n","Requirement already satisfied: tensorboard-data-server<0.7.0,>=0.6.0 in /usr/local/lib/python3.7/dist-packages (from tensorboard~=2.6->tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (0.6.1)\n","Requirement already satisfied: markdown>=2.6.8 in /usr/local/lib/python3.7/dist-packages (from tensorboard~=2.6->tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (3.3.6)\n","Requirement already satisfied: google-auth-oauthlib<0.5,>=0.4.1 in /usr/local/lib/python3.7/dist-packages (from tensorboard~=2.6->tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (0.4.6)\n","Requirement already satisfied: setuptools>=41.0.0 in /usr/local/lib/python3.7/dist-packages (from tensorboard~=2.6->tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (57.4.0)\n","Requirement already satisfied: google-auth<3,>=1.6.3 in /usr/local/lib/python3.7/dist-packages (from tensorboard~=2.6->tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (1.35.0)\n","Requirement already satisfied: cachetools<5.0,>=2.0.0 in /usr/local/lib/python3.7/dist-packages (from google-auth<3,>=1.6.3->tensorboard~=2.6->tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (4.2.4)\n","Requirement already satisfied: pyasn1-modules>=0.2.1 in /usr/local/lib/python3.7/dist-packages (from google-auth<3,>=1.6.3->tensorboard~=2.6->tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (0.2.8)\n","Requirement already satisfied: rsa<5,>=3.1.4 in /usr/local/lib/python3.7/dist-packages (from google-auth<3,>=1.6.3->tensorboard~=2.6->tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (4.8)\n","Requirement already satisfied: requests-oauthlib>=0.7.0 in /usr/local/lib/python3.7/dist-packages (from google-auth-oauthlib<0.5,>=0.4.1->tensorboard~=2.6->tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (1.3.0)\n","Requirement already satisfied: importlib-metadata>=4.4 in /usr/local/lib/python3.7/dist-packages (from markdown>=2.6.8->tensorboard~=2.6->tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (4.10.0)\n","Requirement already satisfied: zipp>=0.5 in /usr/local/lib/python3.7/dist-packages (from importlib-metadata>=4.4->markdown>=2.6.8->tensorboard~=2.6->tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (3.7.0)\n","Requirement already satisfied: pyasn1<0.5.0,>=0.4.6 in /usr/local/lib/python3.7/dist-packages (from pyasn1-modules>=0.2.1->google-auth<3,>=1.6.3->tensorboard~=2.6->tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (0.4.8)\n","Requirement already satisfied: idna<3,>=2.5 in /usr/local/lib/python3.7/dist-packages (from requests<3,>=2.21.0->tensorboard~=2.6->tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (2.10)\n","Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /usr/local/lib/python3.7/dist-packages (from requests<3,>=2.21.0->tensorboard~=2.6->tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (1.24.3)\n","Requirement already satisfied: chardet<4,>=3.0.2 in /usr/local/lib/python3.7/dist-packages (from requests<3,>=2.21.0->tensorboard~=2.6->tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (3.0.4)\n","Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.7/dist-packages (from requests<3,>=2.21.0->tensorboard~=2.6->tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (2021.10.8)\n","Requirement already satisfied: oauthlib>=3.0.0 in /usr/local/lib/python3.7/dist-packages (from requests-oauthlib>=0.7.0->google-auth-oauthlib<0.5,>=0.4.1->tensorboard~=2.6->tensorflow->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 2)) (3.1.1)\n","Requirement already satisfied: pytz>=2017.2 in /usr/local/lib/python3.7/dist-packages (from pandas->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 4)) (2018.9)\n","Requirement already satisfied: python-dateutil>=2.7.3 in /usr/local/lib/python3.7/dist-packages (from pandas->-r /content/drive/MyDrive/Colab_Notebooks/L2X/requirements.txt (line 4)) (2.8.2)\n"]}]},{"cell_type":"code","execution_count":15,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"OmPTyPHt3z_m","executionInfo":{"status":"ok","timestamp":1642518850275,"user_tz":-60,"elapsed":8782,"user":{"displayName":"Francisco Pereira","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14Gju-L5-ouxVHtbPGNLO-WBYoOaPx1t8TQtxAGb4kA=s64","userId":"16300594535934500651"}},"outputId":"a72520ca-af84-46d6-ebcf-fddda56f335c"},"outputs":[{"output_type":"stream","name":"stdout","text":["2022-01-18 15:14:05.824608: W tensorflow/core/common_runtime/gpu/gpu_bfc_allocator.cc:39] Overriding allow_growth setting because the TF_FORCE_GPU_ALLOW_GROWTH environment variable is set. Original config value was 0.\n","25/25 [==============================] - 3s 22ms/step\n","25/25 [==============================] - 1s 22ms/step\n","The val accuracy is 0.89524\n","The train accuracy is 0.96088\n"]}],"source":["!python /content/drive/MyDrive/Colab_Notebooks/L2X/imdb-word/explain.py --task original"]},{"cell_type":"code","source":["!python /content/drive/MyDrive/Colab_Notebooks/L2X/imdb-word/explain.py --task L2X --train"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"sKPc_heM4XoB","executionInfo":{"status":"ok","timestamp":1642517899572,"user_tz":-60,"elapsed":92191,"user":{"displayName":"Francisco Pereira","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14Gju-L5-ouxVHtbPGNLO-WBYoOaPx1t8TQtxAGb4kA=s64","userId":"16300594535934500651"}},"outputId":"4796b03a-012f-4b07-8916-176573f43272"},"execution_count":8,"outputs":[{"output_type":"stream","name":"stdout","text":["Loading dataset...\n","Creating model...\n","2022-01-18 14:56:51.449423: W tensorflow/core/common_runtime/gpu/gpu_bfc_allocator.cc:39] Overriding allow_growth setting because the TF_FORCE_GPU_ALLOW_GROWTH environment variable is set. Original config value was 0.\n","KerasTensor(type_spec=TensorSpec(shape=(None, 400, 1), dtype=tf.float32, name=None), name='conv4_gumbel/BiasAdd:0', description=\"created by layer 'conv4_gumbel'\")\n","----->  <module 'keras.backend' from '/usr/local/lib/python3.7/dist-packages/keras/backend.py'>\n","The train and validation accuracy of the original model is 0.99508 and 0.89088\n","Epoch 1/5\n","----->  <module 'keras.backend' from '/usr/local/lib/python3.7/dist-packages/keras/backend.py'>\n","----->  <module 'keras.backend' from '/usr/local/lib/python3.7/dist-packages/keras/backend.py'>\n","624/625 [============================>.] - ETA: 0s - loss: 0.4246 - acc: 0.7974----->  <module 'keras.backend' from '/usr/local/lib/python3.7/dist-packages/keras/backend.py'>\n","\n","Epoch 00001: val_acc improved from -inf to 0.87804, saving model to /content/drive/MyDrive/Colab_Notebooks/L2X/imdb-word/models/l2x.hdf5\n","/usr/local/lib/python3.7/dist-packages/keras/engine/functional.py:1410: CustomMaskWarning: Custom mask layers require a config and must override get_config. When loading, the custom mask layer must be passed to the custom_objects argument.\n","  layer_config = serialize_layer_fn(layer)\n","625/625 [==============================] - 13s 16ms/step - loss: 0.4241 - acc: 0.7976 - val_loss: 0.3227 - val_acc: 0.8780\n","Epoch 2/5\n","624/625 [============================>.] - ETA: 0s - loss: 0.2927 - acc: 0.8833\n","Epoch 00002: val_acc improved from 0.87804 to 0.88652, saving model to /content/drive/MyDrive/Colab_Notebooks/L2X/imdb-word/models/l2x.hdf5\n","625/625 [==============================] - 13s 20ms/step - loss: 0.2927 - acc: 0.8833 - val_loss: 0.3211 - val_acc: 0.8865\n","Epoch 3/5\n","621/625 [============================>.] - ETA: 0s - loss: 0.2577 - acc: 0.9019\n","Epoch 00003: val_acc improved from 0.88652 to 0.88892, saving model to /content/drive/MyDrive/Colab_Notebooks/L2X/imdb-word/models/l2x.hdf5\n","625/625 [==============================] - 10s 16ms/step - loss: 0.2575 - acc: 0.9019 - val_loss: 0.3291 - val_acc: 0.8889\n","Epoch 4/5\n","624/625 [============================>.] - ETA: 0s - loss: 0.2322 - acc: 0.9139\n","Epoch 00004: val_acc improved from 0.88892 to 0.89456, saving model to /content/drive/MyDrive/Colab_Notebooks/L2X/imdb-word/models/l2x.hdf5\n","625/625 [==============================] - 13s 20ms/step - loss: 0.2322 - acc: 0.9139 - val_loss: 0.3254 - val_acc: 0.8946\n","Epoch 5/5\n","622/625 [============================>.] - ETA: 0s - loss: 0.2164 - acc: 0.9218\n","Epoch 00005: val_acc improved from 0.89456 to 0.89468, saving model to /content/drive/MyDrive/Colab_Notebooks/L2X/imdb-word/models/l2x.hdf5\n","625/625 [==============================] - 13s 20ms/step - loss: 0.2164 - acc: 0.9218 - val_loss: 0.3352 - val_acc: 0.8947\n","Training time is 83.12763619422913\n","625/625 [==============================] - 3s 4ms/step\n","Creating dataset with selected sentences...\n"]}]},{"cell_type":"code","source":["!python /content/drive/MyDrive/Colab_Notebooks/L2X/imdb-word/validate_explanation.py"],"metadata":{"id":"bgXXwniI5KxE","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1642519051008,"user_tz":-60,"elapsed":7898,"user":{"displayName":"Francisco Pereira","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14Gju-L5-ouxVHtbPGNLO-WBYoOaPx1t8TQtxAGb4kA=s64","userId":"16300594535934500651"}},"outputId":"d2cba72a-5780-4d2c-f547-638e1fc10c7f"},"execution_count":19,"outputs":[{"output_type":"stream","name":"stdout","text":["Loading dataset...\n","Creating model...\n","2022-01-18 15:17:27.221730: W tensorflow/core/common_runtime/gpu/gpu_bfc_allocator.cc:39] Overriding allow_growth setting because the TF_FORCE_GPU_ALLOW_GROWTH environment variable is set. Original config value was 0.\n","25/25 [==============================] - 3s 22ms/step\n","The validation accuracy with selected 10 words is 0.9048.\n"]}]},{"cell_type":"code","source":[""],"metadata":{"id":"fY2zL_YHlare","executionInfo":{"status":"ok","timestamp":1642517343151,"user_tz":-60,"elapsed":19,"user":{"displayName":"Francisco Pereira","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14Gju-L5-ouxVHtbPGNLO-WBYoOaPx1t8TQtxAGb4kA=s64","userId":"16300594535934500651"}}},"execution_count":5,"outputs":[]}]}